compiler/qsc_partial_eval/src/lib.rs

```diff
@@ -230,23 +282,146 @@ impl<'a> PartialEvaluator<'a> {
+                    "double binary operation".to_string(),
+                    lhs_expr.span,
+                ))
+            }
+            Value::Var(_lhs_eval_var) => {
+                let lhs_expr = self.get_expr(lhs_expr_id);
+                Err(Error::Unimplemented(
+                    "binary operation with dynamic LHS".to_string(),
+                    lhs_expr.span,
+                ))
+            }
+            _ => {
+                let lhs_expr = self.get_expr(lhs_expr_id);
+                Err(Error::Unexpected(
+                    format!("unsupported LHS value: {lhs_value}"),
+                    lhs_expr.span,
+                ))
+            }
+        }
+    }
+
+    fn eval_bin_op_with_lhs_array_operand(
+        &mut self,
+        bin_op: BinOp,
+        lhs_array: &Rc<Vec<Value>>,
+        rhs_expr_id: ExprId,
+        bin_op_expr_span: Span, // For diagnostic purposes only.
+    ) -> Result<EvalControlFlow, Error> {
+        // Check that the binary operation is currently supported.
+        if matches!(bin_op, BinOp::Eq | BinOp::Neq) {
+            return Err(Error::Unimplemented(
+                "array comparison".to_string(),
+                bin_op_expr_span,
+            ));
+        }
+
+        // The only possible binary operation with array operands at this point is addition.
+        assert!(
+            matches!(bin_op, BinOp::Add),
+            "expected array addition operation, got {bin_op:?}"
+        );
+
+        // Try to evaluate the RHS array expression to get its value.
+        let rhs_control_flow = self.try_eval_expr(rhs_expr_id)?;
+        let EvalControlFlow::Continue(rhs_value) = rhs_control_flow else {
+            let rhs_expr = self.get_expr(rhs_expr_id);
+            return Err(Error::Unexpected(
+                "embedded return in RHS expression".to_string(),
+                rhs_expr.span,
+            ));
+        };
+        let Value::Array(rhs_array) = rhs_value else {
+            panic!("expected array value from RHS expression");
+        };
+
+        // Concatenate the arrays.
+        let concatenated_array: Vec<Value> =
+            lhs_array.iter().chain(rhs_array.iter()).cloned().collect();
+        let array_value = Value::Array(concatenated_array.into());
+        Ok(EvalControlFlow::Continue(array_value))
+    }
+
     fn eval_classical_expr(&mut self, expr_id: ExprId) -> Result<EvalControlFlow, Error> {
+        // This is a workaround since evaluating sections of the execution graph that correspond to local variable
+        // expressions does not yield the expected results.
+        let expr = self.get_expr(expr_id);
+        if let ExprKind::Var(Res::Local(local_var_id), _) = &expr.kind {
+            let value = self
+                .eval_context
+                .get_current_scope()
+                .get_classical_local_value(*local_var_id);
+            return Ok(EvalControlFlow::Continue(value.clone()));
+        }
```

Comment:
We should talk about this on Monday.

Reply:
Is there a test case associated with this? I commented it out and everything still passes, so maybe it's not needed (or there is a test missing)?

Reply:
Added some more test cases that without the fix or my initial workaround should fail (now they don't 😄)